src/Nethermind/Nethermind.Specs/ChainSpecStyle/ChainSpecParametersProvider.cs

```diff
@@ -0,0 +1,99 @@
+// SPDX-FileCopyrightText: 2024 Demerzel Solutions Limited
+// SPDX-License-Identifier: LGPL-3.0-only
+
+
+using System.Reflection;
+using System.Text.Json;
+using Nethermind.Config;
+
+namespace Nethermind.Specs.ChainSpecStyle;
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+using Nethermind.Core;
+
+public class ChainSpecParametersProvider : IChainSpecParametersProvider
+{
+    // TODO: test that all IChainSpecEngineParameters have this suffix
+    private const string EngineParamsSuffix = "ChainSpecEngineParameters";
+
+    private readonly Dictionary<string, JsonElement> _chainSpecParameters =
+        new(StringComparer.InvariantCultureIgnoreCase);
+
+    private readonly ConcurrentDictionary<Type, IChainSpecEngineParameters> _instances = new();
+    public string SealEngineType { get; }
+
+    public ChainSpecParametersProvider(Dictionary<string, JsonElement> engineParameters)
+    {
+        foreach (var parameters in engineParameters)
+        {
+            _chainSpecParameters[parameters.Key] = parameters.Value;
+        }
+
+        InitializeInstances();
+        SealEngineType = CalculateSealEngineType();
+    }
+
+    string CalculateSealEngineType()
+    {
+        string? result = null;
+        foreach (IChainSpecEngineParameters item in _instances.Values)
+        {
+            if (item.SealEngineType is not null)
+            {
+                if (result is not null)
+                {
+                    throw new InvalidOperationException("Multiple seal engines in chain spec");
+                }
+
+                result = item.SealEngineType;
+            }
+        }
+
+        if (result is null)
+        {
+            throw new InvalidOperationException("No seal engine in chain spec");
+        }
+
+        return result;
+    }
+
+    private void InitializeInstances()
+    {
+        Type type = typeof(IChainSpecEngineParameters);
+        IEnumerable<Type> types = TypeDiscovery.FindNethermindBasedTypes(type).Where(x => x.IsClass);
```

Comment:
Do we want auto-discovery or explicit registration?

Reply:
In the Transaction type refactoring we've been working with explicit registration. This should make transitioning to a DI approach easier but makes the immediate code harder since you need to register at the appropriate time, a lot of times directly into `static` fields.

Reply:
Same argument, do as in config. But I don't mind both tbh